### DISCUSSION TITLE
Reverse parsing: from AST to code

### DISCUSSION THREAD

**Discussion by jhrmnn** [textX/textX#326] [Q&A]
Hi, amazing project! I just started using it to parse a small DSL that serves as an input file to a large scientific program. 

I was wondering how much work would it be to add the ability to go back from the parsed AST to the source code. The use case would be to parse code, modify the AST to some degree and write it back to code.

(I assume this functionality does not exist at the moment, I didn't find any mention of it in the docs.)

**Comment by igordejanovic** [comment]
Hi, thanks!
You are right. That feature is not implemented at the moment but I do have a plan to do so at some point.
As I see it, there are two goals that need to be met:
- extend the grammar language to provide formatting information,
- during parsing collect and keep layouting (whitespaces) on the AST.

The reason for the first goal is that you need a way to layout newly created elements of the model. 
The second goal is needed as you would want to retain formatting of parsed input files and layout only the newly added parts. This is especially important when using text-based version control systems such as git.

**Comment by jhrmnn** [comment]
Great. Would this also require changes to Arpeggio, or can it be implemented purely in TextX?

**Comment by igordejanovic** [comment]
The first point will probably be done on the textX level alone. Second one will definitely need change at the parser level but I think I already have some pieces in place.
As a side note, there is [a new parsing technology](https://github.com/igordejanovic/parglare) I'm working on that should be an additional front-end for textX eventually. A proper handling of whitespaces is done from the beginning to be able to deal with issues like this.

**Comment by dkrikun** [comment]
This could also be helpful in code-generation, if one would like to preserve (some of) whitespace and formatting from the source code

**Comment by ivan-alfonso** [comment]
Hi,

I would like to know if as of today, this feature (reverse parsing: from AST to code) was addressed, or if it is in the plans to be developed soon?

Thanks for this useful tool.

**Comment by igordejanovic** [comment]
Hi. No further work has been done since. While the feature would be useful I don't have time nor strong incentive to work on at the moment. But, if someone has time to give it a try I would be happy to review PR.
